lls/cache.c

```diff
@@ -408,9 +467,14 @@ process_ping_pkts(struct rte_mbuf **pkts, unsigned int num_pkts,
 
 	num_granted_pkts = tb_ratelimit_allow_n(num_pkts, rs);
 
-	for (i = 0; i < num_granted_pkts; i++)
-		(*xmit_reply_fn)(iface, pkts[i]);
-	send_pkts(iface->id, tx_queue, num_granted_pkts, pkts);
+	for (i = 0; i < num_granted_pkts; i++) {
+		bool succ = (*xmit_reply_fn)(iface, pkts[i]);
+		if (succ)
+			ping_pkts[num_ping_reply_pkts++] = pkts[i];
+		else
+			rte_pktmbuf_free(pkts[i]);
+	}
```

Comment:
1. If there are malformed packets at the beginning of `pkts[]`, this code is going to send less than `num_granted_pkts`. A solution is to loop until `num_ping_reply_pkts >= num_granted_pkts`; the `>` is only for robustness.
2. If you keep packets that will be transmitted at the beginning of `pkt[]` and packets that will be freed at the end of `pkt[]`, we don't need `ping_pkts[]` and block the calls to `rte_pktmbuf_free()` for a future update of our DPDK. To accomplish this, whenever we find a malformed packet, make space at the end of `pkt[]` by swapping the malformed packet with the packet there. This can invert the arrival order of some ping packets, but this is not an issue.